File: 2018/days/Day03.cs
namespace _2018.days;

public class Claim
{
    public int Row { get; set; }
    public int Col { get; set; }
    public int Width { get; set; }
    public int Height { get; set; }
    public int Id { get; set; }
}

public static class Day03
{
    public static void Run()
    {
        char[,] grid = new char[1000, 1000];
        for (int i = 0; i < grid.GetLength(0); i++)
            for (int j = 0; j < grid.GetLength(1); j++)
                grid[i, j] = '.';

        Claim[] claims = [.. File.ReadAllLines("inputs/day03.txt")
            .Select(line =>
            {
                string[] sp = line.Split(' ');
                int id = int.Parse(sp[0][1..]);
                int[] rowcol = sp[2][..^1].Split(',').Select(num => int.Parse(num)).ToArray();
                int[] widthheight = sp[3].Split('x').Select(num => int.Parse(num)).ToArray();
                return new Claim() { Id = id, Col = rowcol[0], Row = rowcol[1], Width = widthheight[0], Height = widthheight[1] };
            })];

        foreach (var claim in claims)
        {
            for (int row = claim.Row; row < claim.Row + claim.Height; row++)
                for (int col = claim.Col; col < claim.Col + claim.Width; col++)
                    grid[row, col] = grid[row, col] == '.' ? '@' : 'x';
        }

        int overlaps = 0;
        for (int row = 0; row < grid.GetLength(0); row++)
            for (int col = 0; col < grid.GetLength(1); col++)
                if (grid[row, col] == 'x') overlaps++;
        System.Console.WriteLine($"Part 1: {overlaps}");

        foreach (var claim in claims)
        {
            for (int row = claim.Row; row < claim.Row + claim.Height; row++)
                for (int col = claim.Col; col < claim.Col + claim.Width; col++)
                    if (grid[row, col] == 'x') goto End;
            Console.WriteLine(claim.Id); 
            break;
            End: {}
        }
    }
}

File: 2018/day03.js
import fs from 'fs';

console.time();

const claims = fs
  .readFileSync("inputs/day03.txt", "utf-8")
  .split("\n")
  .map((line) => {
    let sp = line.split(' ');
    let pos = sp[2].slice(0, sp[2].length - 1).split(',').map(num => parseInt(num));
    let dim = sp[3].split('x').map(num => parseInt(num));
    return {id: sp[0], pos, dim};
  });

const grid = makeGrid(1000);
for (let claim of claims) {
  for (let i = claim.pos[0]; i < claim.pos[0] + claim.dim[0]; i++) {
    for (let j = claim.pos[1]; j < claim.pos[1] + claim.dim[1]; j++) {
      grid[i][j] = grid[i][j] === '.' ? '!' : 'x';
    }
  }
}

let overlap = 0;
for (let i = 0; i < grid.length; i++) {
  for (let j = 0; j < grid.length; j++) {
    if (grid[i][j] === 'x') overlap++;
  }
}
console.log(`Part 1: ${overlap}`);

for (let claim of claims) {
  if (!overlaps(claim, grid)) {
    console.log(`Part 2: ${claim.id}`);
    break;
  }
}

console.timeEnd();

function overlaps(claim, grid) {
  for (let i = claim.pos[0]; i < claim.pos[0] + claim.dim[0]; i++) {
    for (let j = claim.pos[1]; j < claim.pos[1] + claim.dim[1]; j++) {
      if (grid[i][j] === 'x') return true;
    }
  }
  return false;
}

function makeGrid(dim) {
  let grid = [];
  for (let i = 0; i < dim; i++) {
    let row = [];
    for (let j = 0; j < dim; j++) {
      row.push(".");
    }
    grid.push(row);
  }
  return grid;
}
